tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_thunk.cc

```diff
@@ -30,50 +31,45 @@ namespace gpu {
 
 namespace dnn = se::dnn;
 
-static std::pair<dnn::BatchDescriptor /*input_desc*/,
-                 dnn::BatchDescriptor /*scale_offset_desc*/>
-MakeDescriptors(const Shape& shape, int64 feature_index) {
-  std::vector<int64> logical_to_physical =
-      LayoutUtil::MakeLogicalToPhysical(shape.layout());
-
-  auto physical_dim_size = [&](int64 physical_dim) {
-    return shape.dimensions(LayoutUtil::Major(shape.layout(), physical_dim));
-  };
-
-  // Batchnorm only cares about the location of the depth (aka "feature") dim.
-  // The other dims are all treated the same.  Thus we can use the kBatchDepthYX
-  // cudnn layout for any XLA shape+layout, even XLA shapes that don't have
-  // exactly 4 dimensions: We put everything that comes before the feature dim
-  // into "batch", and everything that comes after the feature dim into "Y".
-  int64 batch_size = 1;
-  int64 y_size = 1;
-  int64 physical_dim;
-  for (physical_dim = 0; physical_dim != logical_to_physical[feature_index];
-       ++physical_dim) {
-    CHECK_LT(physical_dim, shape.dimensions_size());
-    batch_size *= physical_dim_size(physical_dim);
+namespace {
+void CheckInputOutputPrimitivetypeValid(const HloInstruction* hlo) {
+  // Check Inputs.
+  int64 num_operands = hlo->operand_count();
+  PrimitiveType operand_dtype = hlo->operand(0)->shape().element_type();
+  CHECK(operand_dtype == F16 || operand_dtype == F32) << "Not yet implemented";
+
+  for (int i = 1; i < num_operands - 2; i++) {
```

Comment:
It might be more readable (although also more verbose) to just switch on the various kinds of batchnorms to make the check more obvious.  E.g. we want to easily read out the supported configs for  forward inference batchnorm.

Reply:
I can do that but do think that is absolutely necessary? Using `Switch-Case` would make it quite verbose and require replication of similar conditions in each case. Also I would still need to carry around the `operand(0)` primitive type anyway to compare within each case. I can add more comments if you think that can help. Please let me know. I can convert this to `Switch-Case` if you that would be helpful.

Reply:
> do think that is absolutely necessary?

No. :) I'm happy to trust your judgement here.

> I can add more comments if you think that can help.

The main readability concern I have is that it isn't obvious from reading the code (for instance) what types we allow for a backward batchnorm op.  If you can refactor the code or add comments to make this obvious some other way that's fine by me.

Reply:
Does this look OK for a description comment in `CheckInputOutputPrimitivetypeIsValid`
 ``` // All input and output statistics variables must be F32. Also, the last
  // operand for CudnnBatchNormForwardInference, CudnnBatchNormForwardTraining,
  // and CudnnBatchNormBackward is the feature_index which must be S64. Apart
  // from this, the type constraints are as follows:
  // CudnnBatchNormForwardInference:
  //            operand[0]: (U: {half, float})
  //                out[0]: (U: {half, float})
  // CudnnBatchNormForwardTraining:
  //            operand[0]: (U: {half, float})
  //                out[0]: (U: {half, float})
  // CudnnBatchNormBackward:
  //            operand[0]: (U: {half, float})
  //            operand[4]: (U: {half, float})
  //                out[0]: (U: {half, float})
  // Note that each type 'U' should be the same for each type of batchnorm
  // thunk.

Reply:
I pushed a modified version of the above comment block. Hopefully this works. Please take a look.